designs/PackageReference-Extern-Alias.md

```diff
@@ -0,0 +1,164 @@
+
+In the csproj the specification is: 
+
+```xml
+<ItemGroup>
+    <PackageReference Include="NuGet.Contoso.Library.Signed"  Version="5.6.0" Aliases="signed" />
+</ItemGroup>
+```
+
+In the targets section of the assets file, where the compile items are listed out: 
+```json
+  "targets": {
+    ".NETCore,Version=v.3.0": {
+      "NuGet.Contoso.Library.Signed/5.6.0": {
+        "type": "package",
+        "compile": {
+          "lib/netstandard2.0/NuGet.Contoso.Library.Signed.dll": {
+            "Aliases" : "signed"
+          }
+        },
+        "runtime": {
+          "lib/netstandard2.0/NuGet.Contoso.Library.Signed.dll": {}
+        }
+      }
+    },
+```
+
+Note that only the compile items have the additional properties. 
+
+On the library side there are no changes. 
+
+The (LockFileItem)[https://github.com/NuGet/NuGet.Client/blob/4fef99532f4022504feec5f68c8501cbeadd3aed/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs] type which represents an element in the compile list already has a collection of properties. 
+
+Specifically:
+
+```cs
+  public string Path { get; }
+
+  public IDictionary<string, string> Properties { get; } = new Dictionary<string, string>();
+```
+
+There is no need for a change here. 
+
+The value provided in the Aliases attribute will be passed through as far as NuGet is concerned. NuGet will not validate that the value provided is a valid alias, that will be done at build time. 
+
+
+The implementation of this feature spans multiple components. 
+Specifically the work items as follows: 
+
+* NuGet.Client makes the changes for processing the new metadata and adding to the project.assets.json
+* The build tasks on (.NET Core SDK side)[https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs] consume these changes
+* The build tasks on the (non-SDK based PackageReference)[https://github.com/dotnet/NuGet.BuildTasks/blob/master/src/Microsoft.NuGet.Build.Tasks/ResolveNuGetPackageAssets.cs] consume these changes.
+* Nomination updates on (project-system)[https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Dependencies/PackageReference.xaml] side.
+
+## Future Work
+
+* Package metadata for transitive package references will be considered in the future. For now the recommendation is to elevate that PackageReference to a direct dependency. 
+
+## Open Questions
+
+* Will create follow ups for the partner issues once this design has been reviewed.
+* In which version will this functionality be ready?
+
+## Considerations
+
+In addition to the proposed approach, 2 of other solutions were considered. 
+
+### Do nothing - Recommend the customers use the custom target workaround
+
+The workaround that allows per assembly granularity is the following:
+
+```xml
+  <Target Name="AddCustomAliases" BeforeTargets="FindReferenceAssembliesForReferences;ResolveReferences">

```

Comment:
What would happen if NuGet only makes the changes to inject this target in PackageReference world, in a way that users don't need to add this target in their project files, instead of adding new elements to project.assets.json?

As far as I understand, in packages.config world, because all references are added to the project file, it is easier to customize it to add the exte

Reply:
It's not really something we want users depending on. It's not a firm public extensibility point. 

It's an internal build target. Yes, you can technically play it to make your scenarios work, but we don't make guarantees that these things will always work the way they do now.